### PR TITLE
Implementation of standard C/C++ fixed width, minimum width, and maximum width types

### PR DESCRIPTION
**Pull request by jsinglet**
Implementation of standard C/C++ fixed width, minimum width, and maximum width types. Also some modification to the way intmax_t and uintmax_t are organized.

### CODE REVIEW

File: cpp/ql/src/semmle/code/cpp/commons/CommonType.qll
Comment (rdmarsh2) [inline cpp/ql/src/semmle/code/cpp/commons/CommonType.qll:88]:
```suggestion
 * An enum type based on a fixed-width integer type. For instance, `enum e: uint8_t = { a, b };`
```

File: cpp/ql/test/library-tests/types/types/types.cpp
Comment (rdmarsh2) [inline cpp/ql/test/library-tests/types/types/types.cpp:84]: You'll need to declare `std::int8_t` here so the test can be extracted properly.

**Review by rdmarsh2 — COMMENTED** [review]
Thanks! Most of this looks good to me. I've requested a few changes to the doc comments, and you'll also need to autoformat your QL files and fix the failing test before we can merge this.

File: cpp/ql/test/library-tests/types/types/types.cpp
Comment (jsinglet) [reply cpp/ql/test/library-tests/types/types/types.cpp]: @rdmarsh2 -- ack, this was a mistake; that definition isn't supposed to be there.

File: cpp/ql/src/semmle/code/cpp/commons/CommonType.qll
Comment (jbj) [inline cpp/ql/src/semmle/code/cpp/commons/CommonType.qll:69]: We've made a big push to avoid public abstract classes in the base libraries, including this file. See https://github.com/github/codeql-c-team/issues/60 for some motivation and context. Abstract classes are great for the library-internal implementation but painful and dangerous to work with for consumers of the library. For code in the base libraries, the balance is toward reducing surprises for the library's consumers.

Comment (geoffw0) [reply cpp/ql/src/semmle/code/cpp/commons/CommonType.qll:69]: What approach do you recommend here?  This class won't behave correctly in its current form without `abstract`.

Comment (MathiasVP) [reply cpp/ql/src/semmle/code/cpp/commons/CommonType.qll:69]: I'd suggest going with a union approach similar to what we've done here: https://github.com/github/codeql/pull/3940

i.e., make a `TFixedWidthIntegralType` that's a union of all the classes that currently implement the abstract class, and then make a `FixedWidthIntegralType` class that extends `TFixedWidthIntegralType`.

Comment (jsinglet) [reply cpp/ql/src/semmle/code/cpp/commons/CommonType.qll]: @jbj @geoffw0 @MathiasVP -- Please see the latest push to this PR which incorporates a bunch of suggestions from @jbj for resolving this issue. Please let me know so we can come to a consensus. 

Also, I'm failing the changelog check but it doesn't look like it's been updated in a while. Shall I create a new directory and note these changes (if so, what directory/version to add?) or omit the changelog entry?

Comment (jsinglet) [reply cpp/ql/src/semmle/code/cpp/commons/CommonType.qll]: reading @MathiasVP -- I'll update and push again.

Comment (MathiasVP) [reply cpp/ql/src/semmle/code/cpp/commons/CommonType.qll]: We've _changed_ (🥁) the way we do change-notes (which is why that directory hasn't been updated in a while). See an example [here](https://github.com/github/codeql/pull/5859) on how we do it now (specifically, the file `cpp/change-notes/2021-10-05-comparison-with-wider-type.md ` created by that PR).

Comment (jsinglet) [reply cpp/ql/src/semmle/code/cpp/commons/CommonType.qll]: Thanks @MathiasVP!

File: cpp/ql/src/semmle/code/cpp/commons/CommonType.qll
Comment (jbj) [inline cpp/ql/src/semmle/code/cpp/commons/CommonType.qll:79]: This worked!? I'm baffled. Did you know this would work, @MathiasVP? It seems like we can derive from an abstract class without extending it by using `extends T` and `this instanceof T` in combination (without `abstract` it would be redundant).

File: cpp/change-notes/2021-06-10-std-types.md
Comment (jbj) [inline cpp/change-notes/2021-06-10-std-types.md:4]:
```suggestion
* Added a type `FixedWidthEnumType` which describes enums based on a fixed-width integer type. For instance, `enum e: uint8_t = { a, b };`.
```

**Review by jbj — APPROVED** [review]
LGTM modulo a typo fix

File: cpp/ql/src/semmle/code/cpp/commons/CommonType.qll
Comment (MathiasVP) [reply cpp/ql/src/semmle/code/cpp/commons/CommonType.qll:79]: I think this is Pavel's "enshrinkening subtyping" pattern, right? It seems to fit this use-case very well, indeed.

Comment (jbj) [reply cpp/ql/src/semmle/code/cpp/commons/CommonType.qll:79]: Aha. For the record, this pattern was discussed in https://github.slack.com/archives/CP0LHP150/p1606300178023900 (and I was apparently in the discussion).

Comment (jsinglet) [reply cpp/ql/src/semmle/code/cpp/commons/CommonType.qll:79]: The ever-talented Mr. @lcartey showed me this pattern in a coding session where we were examining all of the options for representing this (and also the history of why we avoid it). We compared it against some other approaches, for example, one based around ADTs, but in the end this one seems to strike a nice balance between convenience and maintainability.

**Comment by jbj** [comment]
The tests are running as I'm writing this, but I believe earlier there was an autoformat error from "CPP Language Tests 2 (GitHub Actions)".

Also, the QLDoc Checks are asking you to move (no need to copy) the QLDoc from the private abstract classes to the public ones.

**Review by jbj — APPROVED** [review]

**Review by duongxuantuan154 — COMMENTED** [review]
- [ ]

**Pull request merged**
